tests/templates/test_subroutines.py

```diff
@@ -913,7 +913,7 @@ def test_uccsd_operations(self, s_wires, d_wires, weights, ref_gates):
                 [[0, 1, 2]],
                 [],
                 np.array([1.2, 1, 0, 0]),
-                "BasisState parameter must consist of 0 or 1 integers",
+                "Elements of 'init_state' must be integers",
```

Comment:
This had to be changed because `pennylane.numpy.tensor` objects are accepted types here (no check for the type of the elements contained):
https://github.com/PennyLaneAI/pennylane/blob/cf8fe6400efe75a2b75a0313c38bccdd28cb7071/pennylane/templates/subroutines/uccsd.py#L176

With the unwrapping this changed as we'd have a real `ndarray` instead of a `pennylane.numpy.tensor` object:
```python
from pennylane.templates.utils import check_type
from pennylane import numpy as np
import numpy as original_np

init_state = np.array([1.2, 1, 0, 0])

for i in init_state:
    check_type(
        i,
        [int, np.int64, np.ndarray],
        msg="(First check): Elements of 'init_state' must be integers; got {}".format(init_state),
    )

init_state = original_np.array([1.2, 1, 0, 0])

for i in init_state:
    check_type(
        i,
        [int, np.int64, np.ndarray],
        msg="(Second check): Elements of 'init_state' must be integers; got {}".format(init_state),
    )
```
```
~/xanadu/pennylane/pennylane/templates/utils.py in check_type(element, types, msg)
    182 
    183     if not any([isinstance(element, t) for t in types]):
--> 184         raise ValueError(msg)
    185 
    186 

ValueError: (Second check): Elements of 'init_state' must be integers; got [1.2 1.  0.  0. ]
```

Reply:
@antalszava I have to admit, I don't follow the above comment 🤔 

How come the first check pass, but the second one fails, in the example?

Reply:
This test case tests that the UCCSD template will raise an error when supplied `np.array([1.2, 1, 0, 0])` as the initial state.

Previously:

This test case passed because the expected error was raised by the call on line 206 due to the call to `qml.BasisState`. https://github.com/PennyLaneAI/pennylane/blob/cf8fe6400efe75a2b75a0313c38bccdd28cb7071/pennylane/templates/subroutines/uccsd.py#L206

However, conceptually speaking the check on line 176 using `check_type` (linked in the previous comment) should have actually already raised an error. It is meant to check that elements of `init_state` are integer types. However, since it also allows `np.ndarray` as an accepted type, it passes for single element PennyLane tensor objects because if we received a `np.ndarray` we don't further check the type of the element contained. Therefore, when we passed `np.array([1.2, 1, 0, 0])` as a PennyLane tensor and indexed into it using the `for` loop, this check passes.

The snippet in the previous comment proves that this check indeed passes for a PennyLane tensor, but fails for an original `ndarray`. The following is the output of the snippet by adding `    print('Type of element: ', type(i))` right before the `check_type` call:

```
Type of element:  <class 'pennylane.numpy.tensor.tensor'>
Type of element:  <class 'pennylane.numpy.tensor.tensor'>
Type of element:  <class 'pennylane.numpy.tensor.tensor'>
Type of element:  <class 'pennylane.numpy.tensor.tensor'>
Type of element:  <class 'numpy.float64'>
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-18-70b15445a0e0> in <module>
     20         i,
     21         [int, np.int64, np.ndarray],
---> 22         msg="(Second check): Elements of 'init_state' must be integers; got {}".format(init_state),
     23     )

~/xanadu/pennylane/pennylane/templates/utils.py in check_type(element, types, msg)
    182 
    183     if not any([isinstance(element, t) for t in types]):
--> 184         raise ValueError(msg)
    185 
    186 

ValueError: (Second check): Elements of 'init_state' must be integers; got [1.2 1.  0.  0. ]
```

It might be worth revisiting the edge cases for the check using `check_type`, although I would argue that that is out of scope for this PR.

Reply:
Got it! thanks Antal